### ISSUE TITLE
[BOUNTY $1000] Support scaling in ttnn.mean for reduction over any dimension

### ISSUE THREAD

**Issue by marty1885** [tenstorrent/tt-metal#28644]
## 📝 Background
Currently, `ttnn.mean` uses a single reduce OP call when _any one_ of the following is true:

1. We reduce on _all_ dimensions to return a scalar  
2. We reduce on only _one_ dimension and that dimension is either the last (W) or last but one (H)  
3. We reduce on _two_ dimensions and _both_ are last and last but one (H and W)  

See `single_reduce_op` in `ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp`.

When neither of these conditions are met:
1. _transpose_ the reduction dimension to be the H dimension (-2)
2. Invoke reduction kernel
3. _transpose_ again to negate the operation in (1)
4. Finally, use `ttnn.mul_sfpu` to scale down the values.

For performance reasons, we want to avoid (4) and instead pass the scalar to the reduction kernel invocation in (2).

## 🎯 What Success Looks Like
- Modify `reduce_nd_loop` lambda function to accept a scalar value
- Compute the correct `scalar` according to reduction dimensions
- Plumb through the lambda function to pass this scalar to the invocation of "SUM" kernel
- Confirm correct operation with passing unit tests in `tests/ttnn/unit_tests/operations/reduce/test_reduction.py`
-  Prove performance improvement with before/after measurements for inputs of different sizes

## 💡 Problem to Solve
Avoid post-reduction scaling step for mean; instead, pass the scalar directly to the reduction kernel.

## 🧭 Guidance & Starting Points
- Update `reduce_nd_loop` and related kernel invocation logic.
- Ensure all relevant tests pass.

## 🔎 Possible Approaches
- Refactor to compute and pass scalar at reduction call.
- Update tests and validate correctness/performance.

---

## 📚 Resources
- [Original issue #25548](https://github.com/tenstorrent/tt-metal/issues/25548)

**Comment by yiding** [comment]
Interested. Are there any existing tools and procedures for benchmarking?

**Comment by nsorabaTT** [comment]
> Interested. Are there any existing tools and procedures for benchmarking?

Please see the README here: https://github.com/tenstorrent/tt-perf-report/blob/076a1b39/README.md

We use tracy to profile OPs running on the device. The instructions there should help you generate a CSV that has device runtime apart from other metrics. That is a good starting point for benchmarking.

**Comment by PoKert-TT** [comment]
@marty1885 Hello, our team is using the N300 board, and we are very interested in the entire issue. We can try to work on it.

**Comment by debaa98** [comment]
hey @marty1885 can you assign me in this issue ?

**Comment by marty1885** [comment]
@yiding After viewing the documents. Do you want to proceed taking on the bounty? Please also check our [Bounty Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html).

@PoKert-TT @debaa98 I will be processing the requests in order.

**Comment by yiding** [comment]
Yes, I'll take it.

**Comment by PoKert-TT** [comment]
@marty1885 Okay, thank you for your response. We have modified the code as requested and are conducting some benchmark tests. We are waiting for you to assign the PR.

**Comment by yiding** [comment]
Is there any specifically shaped inputs where this is expected to make a big difference? I looked at 3D and 4D cases and the mul_sfpu takes a very small amount of time compared to the other steps.

E.g. for shape (32,4097,2049) with dim (0, 1) and (0, 2), I get this (each is run 5 times, tested on p150):

```
Before:
    %                              OP Code Joined  Device_Time_Sum_us  Ops_Count
67.28               Reduce (in0:dram_interleaved)            15244.90         20
27.63              FillPad (in0:dram_interleaved)             6261.26         10
 5.00            Transpose (in0:dram_interleaved)             1132.06         20
 0.08 UnaryDeviceOperation (in0:dram_interleaved)               19.20         10


After:
    %                   OP Code Joined  Device_Time_Sum_us  Ops_Count
67.42    Reduce (in0:dram_interleaved)            15298.99         20
27.59   FillPad (in0:dram_interleaved)             6261.26         10
 4.99 Transpose (in0:dram_interleaved)             1133.08         20
```

Test case code:

```
@pytest.mark.parametrize("a", [32])
@pytest.mark.parametrize("h", [4097])
@pytest.mark.parametrize("w", [2049])
@pytest.mark.parametrize("dim", [[0, 1], [0, 2]])
@pytest.mark.parametrize("keepdim", [True])
def test_multidim_mean_perf(device, a, h, w, dim, keepdim):
    import tracy
    torch.manual_seed(0)

    for _ in range(5):
        torch_input_tensor = torch_random((a, h, w), -1, 1, dtype=torch.bfloat16)
        input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)

        tracy.signpost(f"start ttnn.mean {dim} {keepdim}")
        output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=keepdim)
        tracy.signpost(f"end ttnn.mean {dim} {keepdim}")

        output_tensor = ttnn.to_torch(output_tensor)
```

**Comment by yiding** [comment]
Also while I'm here... am I crazy or is the existing axis-by-axis implementation wrong? Right now it's dividing by `input_tensor_arg.logical_volume()` which seems like it's only correct if reducing to a scalar? https://github.com/tenstorrent/tt-metal/blob/c2784e782b1ad2e211708c4e7a13494bdfd2d7e9/ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp#L246

The existing test compares the output with torch.mean using `assert_with_pcc`, which I don't think picks up differences from all the values scaled by a constant factor.

```
>>> import torch
>>> import ttnn
>>> device = ttnn.open_device(device_id=0)
>>> ones = ttnn.from_torch(torch.ones((2,2,2, 2)), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
>>> print(ttnn.mean(ones, dim=0))
ttnn.Tensor([[[ 1.00000,  1.00000],
              [ 1.00000,  1.00000]],

             [[ 1.00000,  1.00000],
              [ 1.00000,  1.00000]]], shape=Shape([2, 2, 2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)

>>> print(ttnn.mean(ones, dim=[2, 3]))
ttnn.Tensor([[ 1.00000,  1.00000],
             [ 1.00000,  1.00000]], shape=Shape([2, 2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)

>>> print(ttnn.mean(ones, dim=[0, 1]))
ttnn.Tensor([[ 0.25000,  0.25000],
             [ 0.25000,  0.25000]], shape=Shape([2, 2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)

>>> print(ttnn.mean(ones, dim=[1, 2]))
ttnn.Tensor([[ 0.25000,  0.25000],
             [ 0.25000,  0.25000]], shape=Shape([2, 2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)

>>> print(ttnn.mean(ones, dim=[0, 1, 2]))
ttnn.Tensor([ 0.50000,  0.50000], shape=Shape([2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)
```

Edit: and while while I'm here the `scalar` parameter is handled incorrectly for this `reduce_nd_loop` for other reduce ops as well. Since reduction is axis-by-axis, the scalar should only be applied once, whereas right now it seems to be applied for every axis. Here it is with `max(..., scalar=2)`

```
>>> print(ttnn.max(ones, dim=[0,1], scalar=2))
ttnn.Tensor([[ 4.00000,  4.00000],
             [ 4.00000,  4.00000]], shape=Shape([2, 2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)
>>> ttnn.max(ones, dim=[0, 1, 2], scalar=2)
ttnn.Tensor([ 8.00000,  8.00000], shape=Shape([2]), dtype=DataType::BFLOAT16, layout=Layout::TILE)
>>> ttnn.max(ones, dim=[0, 1, 2, 3], scalar=2)
ttnn.Tensor(16.00000, shape=Shape([]), dtype=DataType::BFLOAT16, layout=Layout::TILE)
```

**Comment by marty1885** [comment]
c.c. @nsorabaTT

**Comment by nsorabaTT** [comment]
> Is there any specifically shaped inputs where this is expected to make a big difference? I looked at 3D and 4D cases and the mul_sfpu takes a very small amount of time compared to the other steps.
> 
> E.g. for shape (32,4097,2049) with dim (0, 1) and (0, 2), I get this (each is run 5 times, tested on p150):
> 
> ```
> Before:
>     %                              OP Code Joined  Device_Time_Sum_us  Ops_Count
> 67.28               Reduce (in0:dram_interleaved)            15244.90         20
> 27.63              FillPad (in0:dram_interleaved)             6261.26         10
>  5.00            Transpose (in0:dram_interleaved)             1132.06         20
>  0.08 UnaryDeviceOperation (in0:dram_interleaved)               19.20         10
> 
> 
> After:
>     %                   OP Code Joined  Device_Time_Sum_us  Ops_Count
> 67.42    Reduce (in0:dram_interleaved)            15298.99         20
> 27.59   FillPad (in0:dram_interleaved)             6261.26         10
>  4.99 Transpose (in0:dram_interleaved)             1133.08         20
> ```
> 
> Test case code:
> 
> ```
> @pytest.mark.parametrize("a", [32])
> @pytest.mark.parametrize("h", [4097])
> @pytest.mark.parametrize("w", [2049])
> @pytest.mark.parametrize("dim", [[0, 1], [0, 2]])
> @pytest.mark.parametrize("keepdim", [True])
> def test_multidim_mean_perf(device, a, h, w, dim, keepdim):
>     import tracy
>     torch.manual_seed(0)
> 
>     for _ in range(5):
>         torch_input_tensor = torch_random((a, h, w), -1, 1, dtype=torch.bfloat16)
>         input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
> 
>         tracy.signpost(f"start ttnn.mean {dim} {keepdim}")
>         output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=keepdim)
>         tracy.signpost(f"end ttnn.mean {dim} {keepdim}")
> 
>         output_tensor = ttnn.to_torch(output_tensor)
> ```

For the shapes, I expect the performance improvement to scale linearly with the size of the dimensions that are NOT being reduced. So getting rid of the extra `mul_sfpu` call and showing proportional decrease in time w.r.t dimensions will be good.

In your test case, since you are also reducing the dimension 1, the final result is of size [1,1,2049]. And hence, the proportion of time taken by the `mul_sfpu` is lower. Trying using test cases where the reduction ratio (size of input : size of output) is not too high. 


To your next question, the plumbing for scaling is incorrect and yes PCC is not great at catching it. We are in the process of moving to a better measure like looking for all_close and/or measuring maximum ULP digits.
This issue is a subset of a bigget set to fix how we handle the `scale` parameter.

**Comment by yiding** [comment]
PR for some of the scalar fixes and the mul_sfpu removal is ready.

I tested (2,2, 1024, 1024) and (2, 2, 4096, 4096) reducing over dim 0 and 1. The cost of mul_sfpu is still small in this case. Details in PR.

**Comment by nsorabaTT** [comment]
@marty1885 This can be moved to "Completed".

**Comment by marty1885** [comment]
@yiding Can you share your email (or send one to `mchang \at tenstorrent.com`) so I can start the payment? I can't find one on your public profile.

if you worry about sharing personal information on the Internet - you can use my GPG key `B36A9C43923914CE` (under email `mchang \at tenstorrent.com`) and post the encrypted text here. Do so with `echo "<username@email.com>" | gpg --encrypt --recipient "mchang@tenstorrent.com"` after importing my key.

**Comment by yiding** [comment]
Sent you a msg on discord via account yiding.jia , let me know if you prefer email instead.
